cpp/src/arrow/acero/swiss_join.cc

```diff
@@ -473,7 +474,7 @@ Status RowArrayMerge::PrepareForMerge(RowArray* target,
     (*first_target_row_id)[sources.size()] = num_rows;
   }
 
-  if (num_bytes > std::numeric_limits<uint32_t>::max()) {
+  if (is_key_data && num_bytes > std::numeric_limits<uint32_t>::max()) {
```

Comment:
When letting non-key data that is greater than `std::numeric_limits<uint32_t>::max()` bypass this check, the `num_bytes` will underflow to a much smaller value in the `static_cast<uint32_t>` in #486. Then the `target` won't allocate enough space, resulting in segfault when copying data to target.